### ISSUE TITLE
Library FlaUILibrary timeout not working

### ISSUE THREAD

**Issue by djalan** [GDATASoftwareAG/robotframework-flaui#201]
Hello

The timeout setting is not working. It is always 10 seconds if I either increase it or decrease it.

I made a repo @ https://github.com/djalan/robot_flaui_timeout

```robot
*** Settings ***
Library     FlaUILibrary    timeout=3000    screenshot_on_failure=${False}
Library     DateTime


*** Test Cases ***
Please work timeout
    ${now}    Get Current Date
    Log To Console    ${now}
    Wait Until Element Exist    /Window[@Name="it does not matter"]
```

**Comment by JimTempAcc** [comment]
```
def _wait_until_element_exist(self, xpath: str, retries: int):
        """Wait until element exist or timeout occurs.

        Args:
            xpath (String): XPath from element which should be hidden
            retries (Number): Maximum number from retries from wait until

        Raises:
            FlaUiError: If element does not exist.
        """

        timer = 0
        old_timeout = self._timeout
        self._set_timeout(0)

        while timer < retries:

            try:
                self._get_element(xpath)
                self._set_timeout(old_timeout)
                return
            except FlaUiError:
                pass

            time.sleep(1)
            timer += 1

        self._set_timeout(old_timeout)
        raise FlaUiError(FlaUiError.ElementNotExists.format(xpath))
```

In current design, the arg 'retries' decides the wait time in any wait_until_element_xxx method.

@djalan u can increase arg 'retries' as u need in current usage.

@Nepitwin IMO, change 'retries' into 'interval' would be better, what about u.
Sample code like
```
def _wait_until_element_exist(self, xpath: str, interval: float):
        old_timeout = self._timeout
        self._set_timeout(0)
        
        max_time = time.time() * 1000 + self._timeout
        while time.time() * 1000 < max_time:
            try:
                self._get_element(xpath)
                self._set_timeout(old_timeout)
                return
            except FlaUiError:
                pass
            time.sleep(interval)
        
        self._set_timeout(old_timeout)
        raise FlaUiError(FlaUiError.ElementNotExists.format(xpath))
```

**Comment by Nepitwin** [comment]
Thanks for you information.

Yes this operation should be asweel used with a timeout handling like mentioned. We will fix this behavior.

**Comment by noubar** [comment]
According to me it is better to not mix the general timeout and wait until keywords,

It is better to have the "wait until" keyword series independent from general timeout and they should have their own retry times and timeouts.

**Comment by Nepitwin** [comment]
Next week i will implement a solution for this use case.

**Comment by djalan** [comment]
> According to me it is better to not mix the general timeout and wait until keywords,
> 
> It is better to have the "wait until" keyword series independent from general timeout and they should have their own retry times and timeouts.

I do not understand why we need a second default timeout for keywords branded as "Wait". This will confuse the user for sure.

Furthermore, it is not pythonic to use milliseconds as you propose. Python only has one "sleep" function and it takes a float as a parameter.

```python
time.sleep(1.1)    # 1100 ms
time.sleep(0.5)    # 500 ms
```

**Comment by noubar** [comment]
@djalan 
Because if you have wait until keyword
then you need to have a loop
The loop should have max times of try and timeout in ms to wait in between tries.
As @JimTempAcc has written in his code example.

Then the question is where do you want to use the general timeout in the loop? as absolute maximal timeout?

then

example:

You want to limit all wait until keywords to the given max timeout. even if the user wants to wait 4 minutes for some loading in one place in all the test suite. Then the user needs to set the general timeout to 4 mins. Then every click failure will wait at least for 4 mins until the keyword fails. Which will result to very slow test runs. This is why the general timeout and wait until timeout should be independent from each other.

**Comment by noubar** [comment]
The general timeout is more about how fast is your system you are testing and how responsive the ui you are testing. To have stable tests.

Faster Cpu and UI means smaller timeouts, slower cpu or UI means bigger general timeout.

**Comment by djalan** [comment]
@noubar Hello!

TLDR: I would stop using the number of retries in the loop and let it be determined by the maximum timeout and the interval. I would use two different timeouts.

I was suggesting to put the default timeout as the default value of the parameter in the function.

```python
def _wait_until_element_exist(self, xpath: str, interval_secs: float=0.2, timeout_secs: float=DEFAULT_TIMEOUT):
```

>  if the user wants to wait 4 minutes for some loading in one place in all the test suite. Then the user needs to set the general timeout to 4 mins.

Why would I change the general timeout to 4 minutes when I can use `wait_until_element_exist` and its parameter.

```python
wait_until_element_exist("//xpath/will/appear/in/four/minutes", timeout=4*60)
```

> The loop should have max times of try

I don't think that a *max times of try* is relevant; I have never seen this in the past in a few different automation libraries. It should be calculated from the interval between each try and the maximum waiting time.

```python
wait_until_element_exist("//xpath", interval=5 timeout=10)    # check twice in 10 sec
wait_until_element_exist("//xpath", interval=1 timeout=10)    # check once every sec for 10 sec
wait_until_element_exist("//xpath", interval=0.25 timeout=10)    # check four times per second for 10 sec
```

But... note that I like your comment regarding the responsiveness of the UI and the PC. For instance, we might need a general timeout of only 1 sec for clicks and things like that. It would not really make sense to use `wait_until_element_exist` to wait for only an extra second (if using the general timeout there); the user probably need a lot more. Something like 5 sec would make more sense. But I prefer a product of the general timeout.

```python
def _wait_until_element_exist(self, xpath: str, interval_secs: float=1.0, timeout_secs: float=GENERAL * 3):
```

general 1 sec --> wait_until_element_exist gives an extra 3 secs
general 10 sec --> wait_until_element_exist gives an extra 30 secs
A flat number like 5 secs is not realistic if the general timeout is already a big 10 sec. That's why I like a multiplication.

**Comment by noubar** [comment]
@djalan 
Very True
We are defending the exact same thing

You can alternatively use the following in your tests until the wait until keywords here are fixed

```

Wait Until Keyword Succeeds    2 min   1s   Element Should Exist   <xpath>
or 
Wait Until Keyword Succeeds    10x   1s   Element Should Exist   <xpath>
```

https://robotframework.org/robotframework/2.6.1/libraries/BuiltIn.html

Look more about The keyword in builtin library of robot
So you can time it to exactly what you want by combining the two keywords

And Element should Exist alone works 100% accuracy with general timeout

**Comment by Nepitwin** [comment]
### Changed

- Wait keywords use Robot Framework time format (`timeout` and `retry_interval`) like BuiltIn.Wait Until Keyword Succeeds
  - `retries` argument is removed from element Wait keywords
  - Bare numbers are seconds, not retry counts or milliseconds
  - Default wait is 10 seconds with a 1 second retry interval
  - Application Wait keywords no longer wait infinitely when timeout is omitted
  - Wait Until Element Is Offscreen raises `Element is visible` when the timeout expires
  - Library import `timeout` and keywords Set Retry Timeout / Reset Retry Timeout are unchanged and still used for delayed find-element retries

  #### Wait Until Element Exist

  Before:

  ```robotframework
  Wait Until Element Exist    ${XPATH}
  Wait Until Element Exist    ${XPATH}    10
  Wait Until Element Exist    ${XPATH}    1
  Wait Until Element Exist    ${XPATH}    10    Custom error
  ```

  Afterwards:

  ```robotframework
  Wait Until Element Exist    ${XPATH}
  Wait Until Element Exist    ${XPATH}    10s
  Wait Until Element Exist    ${XPATH}    1s
  Wait Until Element Exist    ${XPATH}    1000ms    200ms
  Wait Until Element Exist    ${XPATH}    10s    1s    msg=Custom error
  ```

  #### Wait Until Element Does Not Exist

  Before:

  ```robotframework
  Wait Until Element Does Not Exist    ${XPATH}
  Wait Until Element Does Not Exist    ${XPATH}    10
  Wait Until Element Does Not Exist    ${XPATH}    1
  Wait Until Element Does Not Exist    ${XPATH}    10    Custom error
  ```

  Afterwards:

  ```robotframework
  Wait Until Element Does Not Exist    ${XPATH}
  Wait Until Element Does Not Exist    ${XPATH}    10s
  Wait Until Element Does Not Exist    ${XPATH}    1s
  Wait Until Element Does Not Exist    ${XPATH}    1000ms    200ms
  Wait Until Element Does Not Exist    ${XPATH}    10s    1s    msg=Custom error
  ```

  #### Wait Until Element Is Enabled

  Before:

  ```robotframework
  Wait Until Element Is Enabled    ${XPATH}
  Wait Until Element Is Enabled    ${XPATH}    10
  Wait Until Element Is Enabled    ${XPATH}    1
  Wait Until Element Is Enabled    ${XPATH}    10    Custom error
  ```

  Afterwards:

  ```robotframework
  Wait Until Element Is Enabled    ${XPATH}
  Wait Until Element Is Enabled    ${XPATH}    10s
  Wait Until Element Is Enabled    ${XPATH}    1s
  Wait Until Element Is Enabled    ${XPATH}    1000ms    200ms
  Wait Until Element Is Enabled    ${XPATH}    10s    1s    msg=Custom error
  ```

  #### Wait Until Element Is Offscreen

  Before:

  ```robotframework
  Wait Until Element Is Offscreen    ${XPATH}
  Wait Until Element Is Offscreen    ${XPATH}    10
  Wait Until Element Is Offscreen    ${XPATH}    1
  Wait Until Element Is Offscreen    ${XPATH}    10    Custom error
  ```

  Afterwards:

  ```robotframework
  Wait Until Element Is Offscreen    ${XPATH}
  Wait Until Element Is Offscreen    ${XPATH}    10s
  Wait Until Element Is Offscreen    ${XPATH}    1s
  Wait Until Element Is Offscreen    ${XPATH}    1000ms    200ms
  Wait Until Element Is Offscreen    ${XPATH}    10s    1s    msg=Custom error
  ```

  #### Wait For Application Handle By Name

  Before:

  ```robotframework
  Wait For Application Handle By Name    WpfApplication
  Wait For Application Handle By Name    WpfApplication    1000
  ```

  Afterwards:

  ```robotframework
  Wait For Application Handle By Name    WpfApplication
  Wait For Application Handle By Name    WpfApplication    1s
  Wait For Application Handle By Name    WpfApplication    1000ms    200ms
  ```

  #### Wait For Application Handle By Pid

  Before:

  ```robotframework
  Wait For Application Handle By Pid    ${PID}
  Wait For Application Handle By Pid    ${PID}    1000
  ```

  Afterwards:

  ```robotframework
  Wait For Application Handle By Pid    ${PID}
  Wait For Application Handle By Pid    ${PID}    1s
  Wait For Application Handle By Pid    ${PID}    1000ms    200ms
  ```

  #### Wait For Application While Busy By Name

  Before:

  ```robotframework
  Wait For Application While Busy By Name    WpfApplication
  Wait For Application While Busy By Name    WpfApplication    1000
  ```

  Afterwards:

  ```robotframework
  Wait For Application While Busy By Name    WpfApplication
  Wait For Application While Busy By Name    WpfApplication    1s
  Wait For Application While Busy By Name    WpfApplication    1000ms    200ms
  ```

  #### Wait For Application While Busy By Pid

  Before:

  ```robotframework
  Wait For Application While Busy By Pid    ${PID}
  Wait For Application While Busy By Pid    ${PID}    1000
  ```

  Afterwards:

  ```robotframework
  Wait For Application While Busy By Pid    ${PID}
  Wait For Application While Busy By Pid    ${PID}    1s
  Wait For Application While Busy By Pid    ${PID}    1000ms    200ms
  ```

**Comment by Nepitwin** [comment]
### Changed

- Wait keywords use the same retry API as BuiltIn.Wait Until Keyword Succeeds (`retry` and `retry_interval`)
  - First extra argument `retry` is a count (`10x`, `5 times`) or a timeout (`10s`)
  - Second extra argument `retry_interval` is the wait after a failed attempt (`1s`, `200ms`)
  - `retries` argument is removed from element Wait keywords
  - Bare numbers for `retry` are seconds, not retry counts
  - Default wait is `10x` with a `1s` retry interval
  - Application Wait keywords no longer wait infinitely when retry is omitted
  - Exhausted retry raises FlaUiError
  - Application Wait keywords no longer return True
  - Wait Until Element Is Offscreen raises `Element is visible` when retry is exhausted
  - Library import `timeout` and keywords Set Retry Timeout / Reset Retry Timeout are unchanged and still used for delayed find-element retries

  #### Wait Until Element Exist

  Before:

  ```robotframework
  Wait Until Element Exist    ${XPATH}
  Wait Until Element Exist    ${XPATH}    10
  Wait Until Element Exist    ${XPATH}    1
  Wait Until Element Exist    ${XPATH}    10    Custom error
  ```

  Afterwards:

  ```robotframework
  Wait Until Element Exist    ${XPATH}
  Wait Until Element Exist    ${XPATH}    10x
  Wait Until Element Exist    ${XPATH}    2x    1s
  Wait Until Element Exist    ${XPATH}    10x    200ms
  Wait Until Element Exist    ${XPATH}    10x    1s    msg=Custom error
  ```

  #### Wait Until Element Does Not Exist

  Before:

  ```robotframework
  Wait Until Element Does Not Exist    ${XPATH}
  Wait Until Element Does Not Exist    ${XPATH}    10
  Wait Until Element Does Not Exist    ${XPATH}    1
  Wait Until Element Does Not Exist    ${XPATH}    10    Custom error
  ```

  Afterwards:

  ```robotframework
  Wait Until Element Does Not Exist    ${XPATH}
  Wait Until Element Does Not Exist    ${XPATH}    10x
  Wait Until Element Does Not Exist    ${XPATH}    2x    1s
  Wait Until Element Does Not Exist    ${XPATH}    10x    200ms
  Wait Until Element Does Not Exist    ${XPATH}    10x    1s    msg=Custom error
  ```

  #### Wait Until Element Is Enabled

  Before:

  ```robotframework
  Wait Until Element Is Enabled    ${XPATH}
  Wait Until Element Is Enabled    ${XPATH}    10
  Wait Until Element Is Enabled    ${XPATH}    1
  Wait Until Element Is Enabled    ${XPATH}    10    Custom error
  ```

  Afterwards:

  ```robotframework
  Wait Until Element Is Enabled    ${XPATH}
  Wait Until Element Is Enabled    ${XPATH}    10x
  Wait Until Element Is Enabled    ${XPATH}    2x    1s
  Wait Until Element Is Enabled    ${XPATH}    10x    200ms
  Wait Until Element Is Enabled    ${XPATH}    10x    1s    msg=Custom error
  ```

  #### Wait Until Element Is Offscreen

  Before:

  ```robotframework
  Wait Until Element Is Offscreen    ${XPATH}
  Wait Until Element Is Offscreen    ${XPATH}    10
  Wait Until Element Is Offscreen    ${XPATH}    1
  Wait Until Element Is Offscreen    ${XPATH}    10    Custom error
  ```

  Afterwards:

  ```robotframework
  Wait Until Element Is Offscreen    ${XPATH}
  Wait Until Element Is Offscreen    ${XPATH}    10x
  Wait Until Element Is Offscreen    ${XPATH}    2x    1s
  Wait Until Element Is Offscreen    ${XPATH}    10x    200ms
  Wait Until Element Is Offscreen    ${XPATH}    10x    1s    msg=Custom error
  ```

  #### Wait For Application Handle By Name

  Before:

  ```robotframework
  Wait For Application Handle By Name    WpfApplication
  Wait For Application Handle By Name    WpfApplication    1000
  ```

  Afterwards:

  ```robotframework
  Wait For Application Handle By Name    WpfApplication
  Wait For Application Handle By Name    WpfApplication    10x    1s
  Wait For Application Handle By Name    WpfApplication    10x    200ms
  ```

  #### Wait For Application Handle By Pid

  Before:

  ```robotframework
  Wait For Application Handle By Pid    ${PID}
  Wait For Application Handle By Pid    ${PID}    1000
  ```

  Afterwards:

  ```robotframework
  Wait For Application Handle By Pid    ${PID}
  Wait For Application Handle By Pid    ${PID}    10x    1s
  Wait For Application Handle By Pid    ${PID}    10x    200ms
  ```

  #### Wait For Application While Busy By Name

  Before:

  ```robotframework
  Wait For Application While Busy By Name    WpfApplication
  Wait For Application While Busy By Name    WpfApplication    1000
  ```

  Afterwards:

  ```robotframework
  Wait For Application While Busy By Name    WpfApplication
  Wait For Application While Busy By Name    WpfApplication    10x    1s
  Wait For Application While Busy By Name    WpfApplication    10x    200ms
  ```

  #### Wait For Application While Busy By Pid

  Before:

  ```robotframework
  Wait For Application While Busy By Pid    ${PID}
  Wait For Application While Busy By Pid    ${PID}    1000
  ```

  Afterwards:

  ```robotframework
  Wait For Application While Busy By Pid    ${PID}
  Wait For Application While Busy By Pid    ${PID}    10x    1s
  Wait For Application While Busy By Pid    ${PID}    10x    200ms
  ```
